packages/protocol/contracts/liquidity/GrandaMento.sol

```diff
@@ -0,0 +1,477 @@
+   * @param proposal The proposal to get the sell token and sell amount for.
+   */
+  function getSellTokenAndSellAmount(ExchangeProposal memory proposal)
+    private
+    view
+    returns (IERC20, uint256)
+  {
+    IERC20 sellToken;
+    uint256 sellAmount;
+    if (proposal.sellCelo) {
+      sellToken = getGoldToken();
+      sellAmount = proposal.sellAmount;
+    } else {
+      address stableToken = proposal.stableToken;
+      sellToken = IERC20(stableToken);
+      // When selling stableToken, the sell amount is stored in units.
+      // Units must be converted to value when refunding.
+      sellAmount = IStableToken(stableToken).unitsToValue(proposal.sellAmount);
+    }
+    // In the event a precision issue from the unit <-> value calculations results
+    // in sellAmount being greater than this contract's balance, set the sellAmount
+    // to the entire balance.
+    // This check should not be necessary for CELO, but is done so regardless
+    // for extra certainty that cancelling an exchange proposal can never fail
+    // if for some reason the CELO balance of this contract is less than the
+    // recorded sell amount.
+    uint256 totalBalance = sellToken.balanceOf(address(this));
+    if (totalBalance < sellAmount) {
+      sellAmount = totalBalance;
+    }
+    return (sellToken, sellAmount);
+  }
+
+  /**
+   * @notice Using the oracle price, charges the spread and calculates the amount of
+   * the asset being bought.
+   * @dev Stable token value amounts are used for the sellAmount, not unit amounts.
+   * Assumes both CELO and the stable token have 18 decimals.
+   * @param stableToken The stableToken involved in the exchange.
+   * @param sellAmount The amount of the sell token being sold.
+   * @param sellCelo Whether CELO is being sold.
+   * @return The amount of the asset being bought.
+   */
+  function getBuyAmount(address stableToken, uint256 sellAmount, bool sellCelo)
+    public
+    view
+    returns (uint256)
+  {
+    // Gets the price of CELO quoted in stableToken.
+    FixidityLib.Fraction memory exchangeRate = getOracleExchangeRate(stableToken);
+    // If stableToken is being sold, instead use the price of stableToken
+    // quoted in CELO.
+    if (!sellCelo) {
+      exchangeRate = exchangeRate.reciprocal();
+    }
+    // The sell amount taking the spread into account, ie:
+    // (1 - spread) * sellAmount
+    FixidityLib.Fraction memory adjustedSellAmount = FixidityLib.fixed1().subtract(spread).multiply(
+      FixidityLib.newFixed(sellAmount)
+    );
+    // Calculate the buy amount:
+    // exchangeRate * adjustedSellAmount
+    return exchangeRate.multiply(adjustedSellAmount).fromFixed();
+  }
+
+  /**
+   * @notice Gets the proposal identifiers of all exchange proposals in the
+   * Proposed or Approved state.
+   * @return An array of active exchange proposals IDs.
+   */
+  function getActiveProposalIds() external view returns (uint256[] memory) {
```

Comment:
Are we using a linked list only to make this function possible? I guess it's not a *huge* complexity cost, but wondering if we could do without it and just use an array.

Reply:
Yeah. I also first thought of an array and to just have functions that would result in the removal of an element (like cancelling or executing an exchange) pass in the index of the element to the array, and to have something like `require(activeProposalIds[index] == proposalId)`. But in the case of a veto from governance, which would involve creating a governance proposal to call a function like `function cancelExchangeProposal(uint256 proposalId, uint256 activeProposalIdArrayIndex)`, it's possible that the active proposal id array would change so that by the time the governance proposal is able to be executed, the index in the proposal is incorrect.

The linkedlist approach is the only one I thought of that'd be O(1) for inserts/removals & not vulnerable to something like that, and while this view function is O(n), I see it being used for off-chain reads and not in a transaction via a call from another smart contract.

Reply:
Do we need a separate data structure for just active proposals? I guess we're hitting the question of whether we should provide convenient view functions or let people manage their knowledge via something like event indexing. Because removing this would simplify the data structures in this contract and remove a dependency on a new library whose usefulness and overhead we're not sure of, I'm personally slightly leaning towards removing the list, assuming this functionality isn't something very necessary.

Reply:
My only hesitation is I don't think there's a great setup on Celo right now for indexing events and exposing resulting info in a super easy way. It's pretty important for people to be able to see the active proposals and realistically at launch the only sorta user friendly way will be the CLI, and the easiest way is what we've done with governance which is just exposing a view function. The only thing that comes to mind is The Graph but I'm not sure if we're at a point where we feel confident in relying upon that for important indexing. I don't know much about The Graph but I'm also not sure how efficient it'll be to read through all create / cancel / exchanged events in history to figure out the current set of active proposals

Reply:
the indices would only change if we reuse them, right?
maybe a compromise is just to always increment the identifier

Reply:
>maybe a compromise is just to always increment the identifier given that we don't anticipate the number of granda mento proposals to grow rapidly?
Not sure I understand what you meant here?